### ISSUE TITLE
[Tensorboard] Problem with subfolders from SummaryWriter

### ISSUE THREAD

**Issue by GeorgKohl** [pytorch/pytorch#32651]
## Bug

Using the `add_scalars` or the `add_hparams` methods from the `SummaryWriter` in `torch.utils.tensorboard` results in a folder structure with subfolders that each contain further "tfevents" files. As a result, Tensorboard loads them as separate runs which very quickly leads to chaos when trying to distinguish different training sessions:

![Tensorboard](https://user-images.githubusercontent.com/16224924/73186483-17aa0a00-4120-11ea-924c-0335d914f87f.png)

## To Reproduce

The following code creates the problematic folder structure:
`from torch.utils.tensorboard import SummaryWriter`
`writer = SummaryWriter('runs/experiment1')`
`writer.add_hparams({"lr": 0.001, "batch": 10}, {"metric/distance": 0.5})`
`writer.add_scalar("training/loss", 0.783, 1)`
`writer.add_scalar("training/loss", 0.837, 2)`
`writer.add_scalar("training/loss", 0.642, 3)`
`writer.add_scalars("training/accuracy", {"A":0.1, "B":0.2, "C":0.3}, 1)`
`writer.add_scalars("training/accuracy", {"A":0.2, "B":0.3, "C":0.4}, 2)`
`writer.add_scalars("training/accuracy", {"A":0.3, "B":0.4, "C":0.5}, 3)`
`writer.close()`

The resulting files and folders are:
runs/experiment1/1580137040.107033/events.out.tfevents.1580137040.cube1.26313.1
runs/experiment1/training_accuracy_A/events.out.tfevents.1580137040.cube1.26313.2
runs/experiment1/training_accuracy_B/events.out.tfevents.1580137040.cube1.26313.3
runs/experiment1/training_accuracy_C/events.out.tfevents.1580137040.cube1.26313.4
runs/experiment1/events.out.tfevents.1580137040.cube1.26313.0

## Expected behavior

Tensorboard should only show one run containing all data provided to a single `SummaryWriter`. This could be done by always writing to the same "tfevents" file or maybe there is a workaround to prevent tensorboard from recursively searching the runs folder?

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.5.1.10
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] mkl                       2018.0.3                      1  
[conda] mkl_fft                   1.0.6            py37h7dd41cf_0  
[conda] mkl_random                1.0.1            py37h4414c95_1  
[conda] pytorch                   1.3.1           py3.7_cuda9.2.148_cudnn7.6.3_0    pytorch
[conda] tensorflow                1.14.0          mkl_py37h45c423b_0  
[conda] tensorflow-base           1.14.0          mkl_py37h7ce6ba3_0  
[conda] torchvision               0.4.2                 py37_cu92    pytorch

**Comment by Rob-Fletcher** [comment]
I am having the same problem. When I log hparams it is appearing in a subfolder in the log directory.

**Comment by lanpa** [comment]
@GeorgKohl  I think this is the default behavior of tensorboard. Please have a look at https://colab.research.google.com/github/tensorflow/tensorboard/blob/master/docs/hyperparameter_tuning_with_hparams.ipynb You can see that the tensorflow's API also creates many cluttered scalar values. You can use the `filter` field in tensorboard to filter out the hparams as a workaround.

**Comment by GeorgKohl** [comment]
@lanpa Thanks for the clarification. For `add_hparams` I can see the value in having multiple sets of parameters for the same model now. I would argue that using the same architecture with different hyperparameters counts as a new run as the resulting model is different. So I would not use `add_hparams` in that way and put each set of parameters into an entirely new run directory instead, but I guess this comes down to personal preference. 

I was planning to use 'add_scalars' to plot multiple lines in one plot, but your hint to https://github.com/lanpa/tensorboardX/issues/519 drew my attention to the 'add_custom_scalars' function that works perfectly for this purpose without creating multiple subfolders. At this point only having two runs (actual model and hparams subfolder) for each model seems easily manageable with the `filter` field.

Of course it would be a great feature to manually toggle between both behaviors (writing to the same folder or a new subfolder for `add_hparams` or even a non-recursive option when opening tensorboard), but this is not crucial.

**Comment by HamsterHuey** [comment]
@GeorgKohl - I am assuming you are living with the current limitation that if you use `add_custom_scalars()`, that all the colors for all lines in the plot will have the same color since Tensorboard only assigns a single color per run? For me, this makes it impossible to use `add_custom_scalars()` since the lines are all identically colored, and the markers for datapoints are so small in comparison to the line thickness in the charts that it is really difficult to tell different lines apart (especially if some are close to each other).

`add_scalars` solves the color issue, but it does make the UX for selecting runs in Tensorboard client pretty painful. Curious if anyone has a better solution?

**Comment by GeorgKohl** [comment]
Yes, I also noticed this issues with the same color in `add_custom_scalars`, especially with a larger number of lines in one plot. This means having more than 2-4 lines in a single plot is not possible in a way that is visually clear. I found it sometimes useful that they have the same color when comparing a multi-line plot across different runs though. But for me, using `add_scalars` instead is simply not an option as I want to select different runs quickly and often.

Apart from an option to switch between different behaviours, I don't think there is a easy solution for this, that works in all cases. But if someone has a good workaround, I'd also like to know. I feel like tensorboard could be so much more useful, if it would provide more flexibility in terms how things are shown and organized.

**Comment by HamsterHuey** [comment]
Thanks for the reply. Yes, I just migrated our internal training framework codebase to Tensorboard from visdom and while there is a lot to like about Tensorboard in comparison to visdom, it is extremely restrictive and has some really problematic issues like this one from 2017 (https://github.com/tensorflow/tensorboard/issues/170) that still hasn't been addressed.

I'm not entirely sure if giving up on visdom for Tensorboard was the right decision, but if a company with FB's resources isn't keen on funding a better version of Tensorboard, I guess we all need to make do with Tensorboard and all the quirks and limitations associated with it.

**Comment by lanpa** [comment]
"the markers for datapoints are so small in comparison to the line thickness"
@HamsterHuey Adding two sliders for the line thickness and marker size seems helpful. What do you think?

**Comment by HamsterHuey** [comment]
@lanpa Most certainly! Not sure if line-style customization is possible. That would also be another nice way around the limitation of a single color per run. I am assuming these would need to be client-side changes made on the Tensorboard app side of things?

**Comment by lanpa** [comment]
Yes, this feature request should be opened in tensorflow/tensorboard.

**Comment by zakajd** [comment]
For anyone else who face issue with `add_hparams` function.
It's due to this line in pytorch code. You can try to open pull request and fix it. 
https://github.com/pytorch/pytorch/blob/8072f0685f5bd9bc8f1e48ef916518fa31a50826/torch/utils/tensorboard/writer.py#L306

Another solution is to inherit from this class and change one particular line.

```
import os
from torch.utils.tensorboard import SummaryWriter
from torch.utils.tensorboard.summary import hparams

class SummaryWriter(SummaryWriter):
    def add_hparams(self, hparam_dict, metric_dict):
        torch._C._log_api_usage_once("tensorboard.logging.add_hparams")
        if type(hparam_dict) is not dict or type(metric_dict) is not dict:
            raise TypeError('hparam_dict and metric_dict should be dictionary.')
        exp, ssi, sei = hparams(hparam_dict, metric_dict)

        logdir = self._get_file_writer().get_logdir()
        
        with SummaryWriter(log_dir=logdir) as w_hp:
            w_hp.file_writer.add_summary(exp)
            w_hp.file_writer.add_summary(ssi)
            w_hp.file_writer.add_summary(sei)
            for k, v in metric_dict.items():
                w_hp.add_scalar(k, v)

```

**Comment by GeorgKohl** [comment]
@zakajd Your solution via inheritance is an excellent workaround. This really helps to reduce clutter since I'm using `add_hparams` only once for each run anyway. Thank you very much for sharing this!

**Comment by RafaelWO** [comment]
@zakajd Thanks for your solution, worked like charm! 👍

**Comment by Manuscrit** [comment]
For me the change was:

`
import os
from torch.utils.tensorboard import SummaryWriter
from torch.utils.tensorboard.summary import hparams
from tensorboardX import SummaryWriter

class CorrectedSummaryWriter(SummaryWriter):
    def add_hparams(self, hparam_dict, metric_dict):
        torch._C._log_api_usage_once("tensorboard.logging.add_hparams")
        if type(hparam_dict) is not dict or type(metric_dict) is not dict:
            raise TypeError('hparam_dict and metric_dict should be dictionary.')
        exp, ssi, sei = hparams(hparam_dict, metric_dict)

#         logdir = self._get_file_writer().get_logdir()
        
#         with SummaryWriter(log_dir=logdir) as w_hp:
        self.file_writer.add_summary(exp)
        self.file_writer.add_summary(ssi)
        self.file_writer.add_summary(sei)
        for k, v in metric_dict.items():
            self.add_scalar(k, v)
`

**Comment by bonlime** [comment]
Just for the reference. Such behavior was there from the beginning https://github.com/pytorch/pytorch/pull/23134. 
Waiting for the hero to make PR and fix it, while using the solution by @zakajd

**Comment by timurlenk07** [comment]
I have not seen this workaround proposed, but if you define `run_name = '.'` (the reference to the actual directory), then it will log into the same directory (at least in my test).

By the way, changing the following line to `if run_name is None:` will also allow empty string to work properly, and then it logs to the same directory (again, at least in my test on Linux). Will prepare a PR for this soon.
https://github.com/pytorch/pytorch/blob/b106b958eb07810137d5f64d888e18d57e31134f/torch/utils/tensorboard/writer.py#L305

**Comment by tarunbhatiaind** [comment]
@zakajd . Thanks for your solution, but in my case no hyparameters are logged at all. I am adding after each epoch. I hope this was not intended from the code. I am using pytorch with tensorboard dev.

**Comment by zakajd** [comment]
@tarunbhatiaind Proposed solution is 1,5 years old. PyTorch code base definitely changed since that time, so by best advice is to look into code for SummaryWriter and understand what's different now.

**Comment by chnlyi** [comment]
I tried this and it worked. 
`from torch.utils.tensorboard.writer import FileWriter
from torch.utils.tensorboard.summary import scalar
class MySummaryWriter(SummaryWriter):
    def add_scalars(self, main_tag, tag_scalar_dict, global_step=None, walltime=None):
        torch._C._log_api_usage_once("tensorboard.logging.add_scalars")
        walltime = time.time() if walltime is None else walltime
        # fw_logdir = self._get_file_writer().get_logdir()
        for tag, scalar_value in tag_scalar_dict.items():
            # fw_tag = fw_logdir + "/" + main_tag.replace("/", "_") + "_" + tag
            fw_tag = main_tag.replace("/", "_") + "_" + tag
            assert self.all_writers is not None
            if fw_tag in self.all_writers.keys():
                fw = self.all_writers[fw_tag]
            else:
                fw = FileWriter(fw_tag, self.max_queue, self.flush_secs,
                                self.filename_suffix)
                self.all_writers[fw_tag] = fw
            if self._check_caffe2_blob(scalar_value):
                from caffe2.python import workspace
                scalar_value = workspace.FetchBlob(scalar_value)
            fw.add_summary(scalar(main_tag, scalar_value),
                           global_step, walltime)`

**Comment by manusimidt** [comment]
I found a simple but effective workaround.
just set the `run_name` in `add_params()` to the **absolute path** of the event file.

Full Example:
```python
import os
from torch.utils.tensorboard import SummaryWriter

log_path = 'runs/experiment1'
writer = SummaryWriter(log_path)

epochs = 5
last_loss = 0

for i in range(epochs):
    loss = 0.5 ** i  # just some dummy value
    writer.add_scalar("training/loss", loss, i)
    last_loss = loss

writer.add_hparams({"epochs": epochs}, {"hparam/loss": last_loss},
                   run_name=os.path.dirname(os.path.realpath(__file__)) + os.sep + log_path) # <- see here

```

This will log the scalars and the hparams into ONE event file thus they will also be displayed in one run.
![image](https://user-images.githubusercontent.com/29599104/185363943-df5070b0-006b-4e43-b40b-b78488f887b9.png)

**Comment by frankplow** [comment]
Setting `run_name` to `"."` or the path to the log directory makes the hparams appear under the same run for me, but it also has an unfortunate side effect: it stops Tensorboard automatically refreshing when new data is written to the events file. Without setting `run_name`, I can refresh Tensorboard in my browser while a training loop is running to get up-to-date metrics, but if I set `run_name`, I have to restart Tensorboard. I see the following error message in the Tensorboard output which I suspect is related:
```
E1208 11:04:13.848479 10767396864 directory_watcher.py:254] File runs/Dec08_11-02-20_air/events.out.tfevents.1702033340.air.79341.0 updated even though the current file is runs/Dec08_11-02-20_air/events.out.tfevents.1702033341.air.79341.1
```

**Comment by JakobStr** [comment]
Is there a good workaround for the add_scalars nowadays? 
Crazy, that this bug still exists today. Makes it very annoying to use tensorboard for logging multiclass precisions.

**Comment by XZLancer** [comment]
> Is there a good workaround for the add_scalars nowadays? Crazy, that this bug still exists today. Makes it very annoying to use tensorboard for logging multiclass precisions.

Totally agree, looking forward to a solution!

**Comment by anordin95** [comment]
+1 on the `add_scalars` issue. 

There is a slight workaround in the second most (not first!) up-voted answer here: https://stackoverflow.com/questions/48951136/plot-multiple-graphs-in-one-plot-using-tensorboard

Though, it has some drawbacks including manual specification and each line is drawn with the same color.
